test/checkoutput/README.md

```diff
@@ -0,0 +1,54 @@
+# checkoutput tool
+
+The checkoutput tool parses a text or markdown that has a command and the expected output sequence, runs the command and checks the output of the command against the expected output. 
+
+**Usage:**
+```
+<tool> -parsefile <path/to/file> -verbose
+
+-verbose is optional
+-parsefile should specify the path of the input markdown file.
+```
+
+The tool has 2 major functions
+
+1. Parsing and capturing the Run/Verify blocks
+2. Run and verify the output of the blocks captured in the previous step
+
+## Parsing
+The parsing is done using an Finite State Machine(FSM). The FSM has 2 states Start and Parsing. The initial state is Start. The parsing logic is summarized in the following table.
+
+| State | Input | Action | Output State
+|------ |-------|--------|-------------|
+| Start | Run Block | Record the command| Parsing|
+| Parsing | Run Block | Record the command| Parsing|
+| Parsing | Verify Block | Record the output | Start|
+
+An alternative is to use a markdown parser but that would help us only in the Action column and it would introduce external dependencies.
+
+### Example
+
+In the following example, the tool will parse the content of the "Run" block, runs the command. The output of the command is then checked against the text in "Expected output".
+
+Run:
+```console
+go run ./staging/src/k8s.io/component-base/logs/example/cmd/logger.go
+```
+Expected output:
+```
+I0605 22:03:07.224293 3228948 logger.go:58] Log using Infof, key: value
+I0605 22:03:07.224378 3228948 logger.go:59] "Log using InfoS" key="value"
+E0605 22:03:07.224393 3228948 logger.go:61] Log using Errorf, err: fail
+E0605 22:03:07.224402 3228948 logger.go:62] "Log using ErrorS" err="fail"
+I0605 22:03:07.224407 3228948 logger.go:64] Log message has been redacted. Log argument #0 contains: [secret-key]
+```
+## Run and Check output
+The command is run , its output captured and compared with the expected output. If a mismatch is seen the tool flags it as a validation failure. Passes are not reported, but will be output using the verbose flag.
+
+## Failure
+
+Multiple Run/Verify blocks can exist and a failure is treated as an individual failure and will not affect subsequent Run/Verify blocks.
+
+# Open Issues
+## stdout and stderr
+Currently we only check validate stdout. We need to come up with a proper mechanism to validate outputs in stdout and stderr as this tool is intended to check logging.
```

Comment:
As a first step, the tool can capture both stdout and stderr in a single stream. Then different test cases can redirect stdout or stderr to /dev/null to test just one output stream in isolation.

But I agree that this makes writing tests cases more verbose. I agree that it would be nice to support testing "Expected stderr" and "Expected stdout" with separate streams. If one or the other is not specified, the output has to be empty.

Reply:
I am currently using a combined output (both stdout and stderr). We will revisit this to see if we need to have a separate streams.